crates/symbolicator-service/src/metrics.rs

```diff
@@ -1,17 +1,342 @@
+                    }
+                }
+            }
+
+            // send all the aggregated "counter like" metrics
+            for (AggregationKey { ty, name, tags }, value) in total_counters.drain() {
+                let _ = write!(
+                    &mut formatted_metric,
+                    "{prefix}{name}:{value}{ty}{formatted_global_tags}"
+                );
+
+                if let Some(tags) = tags {
+                    if formatted_global_tags.is_empty() {
+                        formatted_metric.push_str("|#");
+                    } else {
+                        formatted_metric.push(',');
+                    }
+                    formatted_metric.push_str(&tags);
+                }
+
+                let _ = sink.emit(&formatted_metric);
+
+                formatted_metric.clear();
+            }
+
+            // send all the aggregated "distribution like" metrics
+            // we do this in a batched manner, as we do not actually *aggregate* them,
+            // but still send each value individually.
+            for (AggregationKey { ty, name, tags }, value) in total_distributions.drain() {
+                suffix.push_str(&formatted_global_tags);
+
+                if let Some(tags) = tags {
+                    if formatted_global_tags.is_empty() {
+                        suffix.push_str("|#");
+                    } else {
+                        suffix.push(',');
+                    }
+                    suffix.push_str(&tags);
+                }
+
+                for batch in value.chunks(DISTRIBUTION_BATCH_SIZE) {
+                    formatted_metric.push_str(&prefix);
+                    formatted_metric.push_str(name);
+
+                    for value in batch {
+                        let _ = write!(&mut formatted_metric, ":{value}");
+                    }
+
+                    formatted_metric.push_str(ty);
+                    formatted_metric.push_str(&suffix);
+
+                    let _ = sink.emit(&formatted_metric);
+                    formatted_metric.clear();
+                }
+
+                suffix.clear();
+            }
+        }
+    };
+
+    thread::Builder::new()
+        .name("metrics-aggregator".into())
+        .spawn(thread_fn)
+        .unwrap();
+
+    local_aggregators
+}
+
+impl MetricsWrapper {
+    /// Invokes the provided callback with a mutable reference to a thread-local [`LocalAggregator`].
+    fn with_local_aggregator(&self, f: impl FnOnce(&mut LocalAggregator)) {
+        let mut local_aggregator = self
+            .local_aggregator
+            .get_or(Default::default)
+            .lock()
+            .unwrap();
+        f(&mut local_aggregator)
+    }
+}
+
+impl Deref for MetricsWrapper {
+    type Target = StatsdClient;
+
+    fn deref(&self) -> &Self::Target {
+        &self.statsd_client
+    }
+}
+
+/// The key by which we group/aggregate metrics.
+#[derive(Eq, Ord, PartialEq, PartialOrd, Hash)]
```

Comment:
Is there a reason this doesn't implement `Debug`/`Clone`?

Reply:
It doesn’t need to be.
Though I just derived `Debug` for this and a bunch of other types that depend on it, instead of manually rolling a `Debug` impl.

I believe `Debug` printing it might be useful in tests and assertions.